### PR TITLE
Preliminary integration tests between Jormungandr <-> Network Layer (`networkTip`)

### PR DESCRIPTION
**Pull request by KtorZ**
# Issue Number

#219

# Overview

- [x] I have slightly changed the behavior of `waitForConnection` to also wait for the network tip to become available.
- [x] I have added happy & negative test paths for `networkTip`
- [x] I have fixed a few things along the way, including configuration and API definition (cf commits)

# Comments

### CODE REVIEW

File: lib/core/src/Cardano/Wallet/Network.hs [hunk 1]
```diff
@@ -63,8 +61,6 @@ instance Exception ErrNetworkUnreachable
 data ErrNetworkTip
     = ErrNetworkTipNetworkUnreachable ErrNetworkUnreachable
     | ErrNetworkTipNotFound
-    | ErrNetworkTipBlockNotFound (Hash "BlockHeader")
-    -- ^ The tip-block wasn't found. This would be surprising.
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Network.hs:67]: Not needed. This is in the end, an implementation details which can be seen as `NetworkTipNotFound` from the external perspective.

Comment (Anviking) [reply lib/core/src/Cardano/Wallet/Network.hs:67]: 🤔 They are different though. `ErrNetworkTipNotFound` is _expected_ to be thrown when the http-bridge is started without the chain. `ErrNetworkTipBlockNotFound ` means "the impossible just happened".

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr/Network.hs [hunk 2]
```diff
@@ -122,15 +121,21 @@ mkJormungandrLayer
     :: Manager -> BaseUrl -> JormungandrLayer IO
 mkJormungandrLayer mgr baseUrl = JormungandrLayer
     { getTipId = ExceptT $ do
-        let ctx = safeLink api (Proxy @GetTipId)
-        run cGetTipId >>= defaultHandler ctx
+        run cGetTipId >>= \case
+            Left (FailureResponse e) | responseStatusCode e == status404 ->
+                return $ Left ErrNetworkTipNotFound
```
Comment (Anviking) [inline lib/jormungandr/src/Cardano/Wallet/Jormungandr/Network.hs]: According to the api spec 404 wasn't an possibility. It is still, or just precautionary?

Ah — you removed it again in a later commit

File: lib/core/src/Cardano/Wallet/Network.hs [hunk 3]
```diff
@@ -90,18 +86,18 @@ waitForConnection nw policy = do
...
-        Left (ErrNetworkTipNetworkUnreachable _) -> do
-             TIO.putStrLn "[INFO] waiting for connection to the node..."
-             return True
-        Left _ -> return True
+        Left ErrNetworkTipNotFound ->
+            return True
+        Left (ErrNetworkTipNetworkUnreachable _) ->
+            return True
 
+-- | A default 'RetryPolicy' with a constant delay, but no longer than 20
```
Comment (Anviking) [inline lib/core/src/Cardano/Wallet/Network.hs:97]: Seems confusing to say that the "constant delay" "is limited". Lazy suggestion:
```suggestion
-- | A default 'RetryPolicy' with a constant delay, but retries for no longer than 20
```

File: lib/jormungandr/test/integration/Cardano/LauncherSpec.hs [hunk 4]
```diff
@@ -12,12 +12,15 @@ import Control.Concurrent.Async
...
     ( void )
+import System.Directory
+    ( removePathForcibly )
 import Test.Hspec
     ( Spec, describe, expectationFailure, it )
 
 spec :: Spec
 spec = describe "cardano-wallet-launcher" $ do
     it "Can start launcher against testnet" $ do
+        removePathForcibly "/tmp/cardano-wallet-jormungandr"
```
Comment (Anviking) [inline lib/jormungandr/test/integration/Cardano/LauncherSpec.hs:23]: Should we remove the storage from the Jörmungandr config, to use in-memory instead?

File: lib/jormungandr/test/integration/Cardano/Wallet/Jormungandr/NetworkSpec.hs [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+spec :: Spec
+spec = do
+    let startNode' = startNode url (`waitForConnection` defaultRetryPolicy)
+    describe "Happy Paths" $ beforeAll startNode' $ afterAll killNode $ do
+        it "get network tip" $ \(_, nw) -> do
+            resp <- runExceptT $ networkTip nw
+            resp `shouldSatisfy` isRight
+            let (Right slot) = slotId . snd <$> resp
+            slot `shouldSatisfy` (>= SlotId 0 0)
```
Comment (Anviking) [inline lib/jormungandr/test/integration/Cardano/Wallet/Jormungandr/NetworkSpec.hs:70]: 👍

File: lib/core/src/Cardano/Wallet/Network.hs [hunk 6]
```diff
@@ -90,18 +86,18 @@ waitForConnection nw policy = do
...
+        Right _ ->
             return False
-        Left (ErrNetworkTipNetworkUnreachable _) -> do
-             TIO.putStrLn "[INFO] waiting for connection to the node..."
-             return True
-        Left _ -> return True
+        Left ErrNetworkTipNotFound ->
+            return True
+        Left (ErrNetworkTipNetworkUnreachable _) ->
+            return True
```
Comment (Anviking) [inline lib/core/src/Cardano/Wallet/Network.hs:95]: You said this was a slight change in behaviour — but it isn't (aside from removing the logging)?

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr/Api.hs [hunk 7]
```diff
@@ -87,14 +85,12 @@ type GetBlockDescendantIds
...
-    = "api"
-    :> "v0"
+    = "v0"
     :> "tip"
     :> Get '[Hex] BlockId
 
 type PostSignedTx
-    = "api"
-    :> "v0"
+    = "v0"
```
Comment (Anviking) [inline lib/jormungandr/src/Cardano/Wallet/Jormungandr/Api.hs:93]: Tried to do that at one point, but gave up because I messed it up somehow 🙃 👍

**Review by Anviking — APPROVED** [review]
lgtm — most pressingly wondering if we shouldn't run jormungandr in-memory

I am also (still?) suspicious/confused about `NetworkLayer` error types, but that's not specifically related to this pr

File: lib/core/src/Cardano/Wallet/Network.hs (hunk 1 above)
Comment (KtorZ) [reply lib/core/src/Cardano/Wallet/Network.hs:67]: But they pretty much means the same for us in the end: there's no tip. We won't do anything special when the corresponding block isn't found bit treat that as if there was no tip.

File: lib/core/src/Cardano/Wallet/Network.hs (hunk 6 above)
Comment (KtorZ) [reply lib/core/src/Cardano/Wallet/Network.hs:95]: It now retries until the tip is found. It used to only wait for the connection to be established.

File: lib/jormungandr/test/integration/Cardano/LauncherSpec.hs (hunk 4 above)
Comment (KtorZ) [reply lib/jormungandr/test/integration/Cardano/LauncherSpec.hs:23]: Hmmmm. No. Ideally integration tests reproduce the production setup as close as possible. This means, among othert things, using file based storage.

File: lib/core/src/Cardano/Wallet/Network.hs (hunk 6 above)
Comment (Anviking) [reply lib/core/src/Cardano/Wallet/Network.hs:95]: Previously the there was a `Left _ -> True`:
```haskell
        Right _ -> do
            return False
        Left (ErrNetworkTipNetworkUnreachable _) -> do
             TIO.putStrLn "[INFO] waiting for connection to the node..."
             return True
        Left _ -> return True
```
which I admit was very weird as it was only logging in the `ErrNetworkTipNetworkUnreachable _` case.

But regardless 👍

**Review by paweljakubas — APPROVED** [review]
LGTM! I am also in favour of using only `ErrNetworkTipNotFound` and ignoring `ErrNetworkTipBlockNotFound (Hash "BlockHeader")` provided we know for sure the error I write below is eliminated. To my understand we get header hash of tip of block so having the tip means we have header hash of the tip. Nevertheless, basing on my experience when playing extensively with diffusion layer of cardano-sl, something else can pop out. Very rarely, but quite enough to detect it in the tests I encountered the error when calling `requestTip` : 
```
BlockNetLogicException: DialogUnexpected "peer sent more than one tip"
```
Presumably, in new Shelly haskell node it is addressed but if not, we should remember about it when having obscure errors.

File: lib/core/src/Cardano/Wallet/Network.hs (hunk 6 above)
Comment (KtorZ) [reply lib/core/src/Cardano/Wallet/Network.hs:95]: Hmm.. Weird. I didn't see this case on my local branch. Either I wasn't up-to-date when making the change or, I can't see things right ^^"
Anyway, no major behavioral change then.

**Comment by KtorZ** [comment]
@pawel.jakubas @anviking

This `ErrNetworkTipBlockNotFound` is a nice example of `warning` log severity IMO. Instead of returning a new error type that we are going to treat like the absence of tip anyway, I'd suggest to have a warning log line mentioning this bit still, returning a network tip not found error.

**Comment by KtorZ** [comment]
@Anviking Merging without the comment suggestion (will do in another PR)

**Pull request merged**